src/man/man1/mbabsorption.1
.TH mbabsorption 1 "3 June 2013" "MB-System 5.0" "MB-System 5.0"
.SH NAME
\fBmbabsorption\fP \- calculates the absorption of sound in sea water in dB/km
as a function of frequency, temperature, salinity, sound speed, and depth.


.SH VERSION
Version 5.0

.SH SYNOPSIS
\fBmbabsorption\fP
.br
[--help {-H}
.br
--sound-speed=soundspeed {-Csoundspeed}
.br
--depth=depth {-Ddepth}
.br
--frequency=frequency {-Ffrequency}
.br
--ph=ph {-Pph}
.br
--salinity=salinity {-Ssalinity}
.br
--temperature=temperature {-Ttemperature}
.br
--verbose {-V}]

.SH DESCRIPTION
\fBmbabsorption\fP \- calculates the absorption of sound in sea water in dB/km
 as a function of frequency (kHz), temperature (degrees C), salinity (per mil), 
depth (m), pH, and water sound speed (m/sec). 

 We use the Francois and Garrison equations from:
    Francois, R.E., Garrison, G.R., "Sound absorption based
      on ocean measurements: Part I: Pure water and magnesium
      sulfate contributions", J. Acoust. Soc. Am., 72(3), 
      896-907, 1982.
    Francois, R.E., Garrison, G.R., "Sound absorption based
      on ocean measurements: Part II: Boric acid contribution
      and equation for total absorption", J. Acoust. Soc. Am., 
      72(6), 1879-1890, 1982.

Francois and Garrison [1982] model the sound absorption in
sea water as resulting from contributions from pure water,
magnesium sulfate, and boric acid. The boric acid contribution
is significant below 10 kHz. The equations are:

 	absorption = Boric Acid Contribution
 			+ MbSO4 Contribution
 			+ Pure Water Contribution
 	
.br
Boric Acid Contribution:
 	AlphaB = Ab * Pb * Fb * f**2
 	         \-------------------
 	            f**2 + Fb**2
 	    
 	Ab = 8.86 / c * 10**(0.78 * pH \- 5) (dB/km/kHz)
 	Pb = 1
 	Fb = 2.8 * (S / 35)**0.5 * 10**(4 \- 1245 / Tk) (kHz)
 		
.br
MgSO4 Contribution:
 	AlphaM = Am * Pm * Fm * f**2
 	         \-------------------
 	            f**2 + Fm**2
		
 	Am = 21.44 * S * (1 + 0.025 * T) / c (dB/km/kHZ)
 	Pm = 1 \- 0.000137 * D + 0.0000000062 * D**2
 	Fm = (8.17 * 10**(8 \- 1990 / Tk)) / (1 + 0.0018 * (S \- 35))  (kHz)
 
.br
Pure Water Contribution:
 	AlphaW = Aw * Pw * f**2
  
 	For T <= 20 deg C
 	  Aw = 0.0004397 \- 0.0000259 * T
 	          + 0.000000911 * T**2 \- 0.000000015 * T**3 (dB/km/kHz) 
 	For T > 20 deg C
 	  Aw = 0.0003964 \- 0.00001146 * T
 	          + 0.000000145 * T**2 \- 0.00000000049 * T**3 (dB/km/kHz) 
 	Pw = 1 \- 0.0000383 * D + 0.00000000049 * D**2 
 
.br
Parameters:
 	f = sound frequency (kHz)
 	c = speed of sound (m/s)
 	T = temperature (deg C)
 	Tk = temperature (deg K) = T + 273 (deg K)
 	S = salinity (per mil)
 	D = depth (m)

.br
If the speed of sound is not specified by the user, it will be
calculated from temperature, salinity, and depth using:
.br
 	c =~ 1412 + 3.21 * T + 1.19 * S + 0.0167 * D

.br
Normally, \fBmbabsorption\fP simply prints the resulting absorption
value to stdout. If the \fB\-V\fP option is specified, the output will
include a listing of the parameters used in calculating the absorption.
 
.SH MB-SYSTEM AUTHORSHIP
David W. Caress
.br
  Monterey Bay Aquarium Research Institute
.br
Dale N. Chayes
.br
  Center for Coastal and Ocean Mapping
.br
  University of New Hampshire
.br
Christian do Santos Ferreira
.br
  MARUM - Center for Marine Environmental Sciences
.br
  University of Bremen

.SH OPTIONS
.TP
.B \-\-help \-H
This "help" flag cause the program to print out a description
of its operation and then exit immediately.
.TP
.B \-\-sound\-speed=soundspeed \-Csoundspeed
Specifies the water sound speed in m/sec. If the option is
not used, the water sound speed will be calculated from the
temperature, salinity, and depth.
.TP
.B \-\-depth=depth \-Ddepth
Sets the depth in meters at which the sound absorption should be calculated.
Default: \fIdepth\fP = 0.0 m (i.e. the sea surface).
.TP
.B \-\-frequency=frequency \-Ffrequency
Sets the sound frequency in kHz at which the sound absorption should be calculated.
Default: \fIfrequency\fP = 200 kHz.
.TP
.B \-\-ph=ph \-Pph
Sets the pH value at which the sound absorption should be calculated.
Default: \fIph\fP = 8.
.TP
.B \-\-salinity=salinity \-Ssalinity
Sets the salinity in per mil at which the sound absorption should be calculated.
Default: \fIsalinity\fP = 35 per mil.
.TP
.B \-\-temperature=temperature \-Ttemperature
Sets the temperature in degrees C at which the sound absorption should be calculated.
Default: \fItemperature\fP = 10.0 degrees C.
.TP
.B \-\-verbose \-V
Normally, \fBmbabsorption\fP outputs only a single absorption value
to the stdout stream.  If the
\fB\-V\fP flag is given, then \fBmbabsorption\fP works in a "verbose" mode and
also outputs the parameters used to calculate the absorption.

.SH EXAMPLES
Suppose that one wishes to obtain the water sound absorption for
a 200 kHz multibeam sonar operating at a depth of 1000 meters (on
an AUV, for instance) in water with a salinity of 35 per mil, a pH
of 8, and a temperature of 4 degrees C. The following will suffice:
.br
 	mbabsorption \-D1000 \-F200 \-P8 \-S35 \-T4
.br
The output will be:
 37.692561
 
.br
In order to obtain a more readable result, use the \fB\-V\fP option:
.br
 	mbabsorption \-D1000 \-F200 \-P8 \-S35 \-T4 \-V
.br
The output will be:
.br
 Program MBabsorption
 MB-system Version 5.1.1beta15
 
 Input Parameters:
      Frequency:        200.000000 kHz
      Temperature:      4.000000 deg C
      Salinity:         35.000000 per mil
      Depth:            1000.000000 m
      pH:               8.000000
 Result:
      Sound absorption: 37.692561 dB/km

.SH SEE ALSO
\fBmbsystem\fP(1)

.SH BUGS
None known.


NEWS
-------------------------------------------------------------------------------
MB-SYSTEM NEWS FILE:
-------------------------------------------------------------------------------
-------------------------------------------------------------------------------

In the GNU package standard, the NEWS file lists the history of user-visible
changes to a package. For the the MB-System open source software package for the
processing and display of swath sonar data, this information is included in the
ChangeLog file. Both the NEWS file and the ChangeLog file are located at the top
of the MB-System source code distribution directory structure.

-------------------------------------------------------------------------------


src/utilities/mbabsorption.cc
/*--------------------------------------------------------------------
 *    The MB-system:	mbabsorption.c	2/10/2008
 *
 *    Copyright (c) 2008-2025 by
 *    David W. Caress (caress@mbari.org)
 *      Monterey Bay Aquarium Research Institute
 *      Moss Landing, California, USA
 *    Dale N. Chayes 
 *      Center for Coastal and Ocean Mapping
 *      University of New Hampshire
 *      Durham, New Hampshire, USA
 *    Christian dos Santos Ferreira
 *      MARUM
 *      University of Bremen
 *      Bremen Germany
 *     
 *    MB-System was created by Caress and Chayes in 1992 at the
 *      Lamont-Doherty Earth Observatory
 *      Columbia University
 *      Palisades, NY 10964
 *
 *    See README.md file for copying and redistribution conditions.
 *--------------------------------------------------------------------*/
/*
 * MBabsorption calculates the absorption of sound in sea water in dB/km
 * as a function of frequency, temperature, salinity, sound speed, and depth.
 *
 * We use the Francois and Garrison equations from:
 *     Francois, R.E., Garrison, G.R., "Sound absorption based
 *       on ocean measurements: Part I: Pure water and magnesium
 *       sulfate contributions", J. Acoust. Soc. Am., 72(3),
 *       896-907, 1982.
 *     Francois, R.E., Garrison, G.R., "Sound absorption based
 *       on ocean measurements: Part II: Boric acid contribution
 *       and equation for total absorption", J. Acoust. Soc. Am.,
 *       72(6), 1879-1890, 1982.
 *
 * Francois and Garrison [1982] model the sound absorption in
 * sea water as resulting from contributions from pure water,
 * magnesium sulfate, and boric acid. The boric acid contribution
 * is significant below 10 kHz. The equations are:
 *
 * absorption = Boric Acid Contribution
 * 		+ MbSO4 Contribution
 * 		+ Pure Water Contribution
 *
 * **************************
 *
 * Boric Acid Contribution
 * AlphaB = Ab * Pb * Fb * f**2
 *          -------------------
 *             f**2 + Fb**2
 *
 * Ab = 8.86 / c * 10**(0.78 * pH - 5) (dB/km/kHz)
 * Pb = 1
 * Fb = 2.8 * (S / 35)**0.5 * 10**(4 - 1245 / Tk) (kHz)
 *
 * **************************
 *
 * MgSO4 Contribution
 * AlphaM = Am * Pm * Fm * f**2
 *          -------------------
 *             f**2 + Fm**2
 *
 * Am = 21.44 * S * (1 + 0.025 * T) / c (dB/km/kHZ)
 * Pm = 1 - 0.000137 * D + 0.0000000062 * D**2
 * Fm = (8.17 * 10**(8 - 1990 / Tk)) / (1 + 0.0018 * (S - 35))  (kHz)
 *
 * **************************
 *
 * Pure Water Contribution
 * AlphaW = Aw * Pw * f**2
 *
 * For T <= 20 deg C
 *   Aw = 0.0004397 - 0.0000259 * T
 *           + 0.000000911 * T**2 - 0.000000015 * T**3 (dB/km/kHz)
 * For T > 20 deg C
 *   Aw = 0.0003964 - 0.00001146 * T
 *           + 0.000000145 * T**2 - 0.00000000049 * T**3 (dB/km/kHz)
 * Pw = 1 - 0.0000383 * D + 0.00000000049 * D**2
 *
 * **************************
 *
 * f = sound frequency (kHz)
 * c = speed of sound (m/s)
 *   =~ 1412 + 3.21 * T + 1.19 * S + 0.0167 * D
 * T = temperature (deg C)
 * Tk = temperature (deg K) = T + 273 (deg K)
 * S = salinity (per mil)
 * D = depth (m)
 *
 * **************************
 *
 * Author:	D. W. Caress
 * Date:	February 10, 2008
 *              R/V Zephyr
 *              Hanging out at the channel entrance to La Paz, BCS, MX
 *              helping out as MBARI tries to save the grounded
 *              R/V Western Flyer.
 *              Note: as I was writing this code the Flyer was refloated
 *              and successfully backed off the reef.
 */

#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#ifdef _WIN32
#include "unistd_w.h"
#else
#include <unistd.h>
#endif
#include "mb_define.h"
#include "mb_status.h"

constexpr char program_name[] = "MBabsorption";
constexpr char help_message[] =
    "MBabsorption calculates the absorption of sound in sea water\n"
    "in dB/km as a function of frequency, temperature, salinity,\n"
    "sound speed, pH, and depth.";
constexpr char usage_message[] =
    "mbabsorption\n"
    "\t--depth=depth {-Ddepth}\n"
    "\t--frequency=frequency {-Ffrequency}\n"
    "\t--help {-H}\n"
    "\t--ph=ph {-Pph}\n"
    "\t--salinity=salinity {-Ssalinity}\n"
    "\t--sound-speed=soundspeed {-Csoundspeed}\n"
    "\t--temperature=temperature {-Ttemperature}\n"
    "\t--verbose {-V}\n\n";

/*--------------------------------------------------------------------*/

int main(int argc, char **argv) {
	int verbose = 0;

	double frequency = 200.0;   /* frequency (kHz) */
	double temperature = 10.0; /* temperature (deg C) */
	double salinity = 35.0;    /* salinity (per mil) */
	double soundspeed = 0.0;  /* speed of sound (m/sec) */
	double depth = 0.0;       /* depth (m) */
	double ph = 8.0;          /* pH */

	/* process argument list */
	bool help = false;
	{
		static struct option options[] = {{"verbose", no_argument, nullptr, 0},
		                                  {"help", no_argument, nullptr, 0},
		                                  {"sound-speed", required_argument, nullptr, 0},
		                                  {"depth", required_argument, nullptr, 0},
		                                  {"frequency", required_argument, nullptr, 0},
		                                  {"ph", required_argument, nullptr, 0},
		                                  {"salinity", required_argument, nullptr, 0},
		                                  {"temperature", required_argument, nullptr, 0},
		                                  {nullptr, 0, nullptr, 0}};

		bool errflg = false;
		int c;
		int option_index;
		while ((c = getopt_long(argc, argv, "VvHhC:c:D:d:F:f:P:p:S:s:T:t:", options, &option_index)) != -1)
			switch (c) {
			/* long options all return c=0 */
			case 0:
				if (strcmp("verbose", options[option_index].name) == 0) {
					verbose++;
				}
				else if (strcmp("help", options[option_index].name) == 0) {
					help = true;
				}
				else if (strcmp("sound-speed", options[option_index].name) == 0) {
					sscanf(optarg, "%lf", &soundspeed);
				}
				else if (strcmp("depth", options[option_index].name) == 0) {
					sscanf(optarg, "%lf", &depth);
				}
				else if (strcmp("frequency", options[option_index].name) == 0) {
					sscanf(optarg, "%lf", &frequency);
				}
				else if (strcmp("ph", options[option_index].name) == 0) {
					sscanf(optarg, "%lf", &ph);
				}
				else if (strcmp("salinity", options[option_index].name) == 0) {
					sscanf(optarg, "%lf", &salinity);
				}
				else if (strcmp("temperature", options[option_index].name) == 0) {
					sscanf(optarg, "%lf", &temperature);
				}
				break;
			case 'H':
			case 'h':
				help = true;
				break;
			case 'V':
			case 'v':
				verbose++;
				break;
			case 'C':
			case 'c':
				sscanf(optarg, "%lf", &soundspeed);
				break;
			case 'D':
			case 'd':
				sscanf(optarg, "%lf", &depth);
				break;
			case 'F':
			case 'f':
				sscanf(optarg, "%lf", &frequency);
				break;
			case 'P':
			case 'p':
				sscanf(optarg, "%lf", &ph);
				break;
			case 'S':
			case 's':
				sscanf(optarg, "%lf", &salinity);
				break;
			case 'T':
			case 't':
				sscanf(optarg, "%lf", &temperature);
				break;
			case '?':
				errflg = true;
			}

		if (errflg) {
			fprintf(stderr, "usage: %s\n", usage_message);
			exit(MB_ERROR_BAD_USAGE);
		}
	}

	/* set output stream */
	FILE *outfp;
	if (verbose <= 1)
		outfp = stdout;
	else
		outfp = stderr;

	if (verbose == 1 || help) {
		fprintf(outfp, "\nProgram %s\n", program_name);
		fprintf(outfp, "MB-system Version %s\n", MB_VERSION);
	}

	if (verbose >= 2) {
		fprintf(outfp, "\ndbg2  Program <%s>\n", program_name);
		fprintf(outfp, "dbg2  MB-system Version %s\n", MB_VERSION);
		fprintf(outfp, "dbg2  Control Parameters:\n");
		fprintf(outfp, "dbg2       verbose:    %d\n", verbose);
		fprintf(outfp, "dbg2       help:       %d\n", help);
		fprintf(outfp, "dbg2       frequency:  %f\n", frequency);
		fprintf(outfp, "dbg2       temperature:%f\n", temperature);
		fprintf(outfp, "dbg2       salinity:   %f\n", salinity);
		fprintf(outfp, "dbg2       soundspeed: %f\n", soundspeed);
		fprintf(outfp, "dbg2       depth:      %f\n", depth);
		fprintf(outfp, "dbg2       ph:         %f\n", ph);
	}

	/* if help desired then print it and exit */
	if (help) {
		fprintf(outfp, "\n%s\n", help_message);
		fprintf(outfp, "\nusage: %s\n", usage_message);
		exit(MB_ERROR_NO_ERROR);
	}

	/* call function to calculate absorption */
	double absorption;  /* absorption (dB/km) */
	double density;     /* kg/m3 */

	int error = MB_ERROR_NO_ERROR;
	int status = mb_absorption(verbose, frequency, temperature, salinity, depth, ph, soundspeed, &absorption, &error);
	const double pressure = 1.006 * depth;  /* depth (m) */
	status &= mb_seabird_density(verbose, salinity, temperature, pressure, &density, &error);

	if (verbose > 0) {
		fprintf(outfp, "\nProgram <%s>\n", program_name);
		fprintf(outfp, "MB-system Version %s\n", MB_VERSION);
		fprintf(outfp, "Input Parameters:\n");
		fprintf(outfp, "     Frequency:        %f kHz\n", frequency);
		fprintf(outfp, "     Temperature:      %f deg C\n", temperature);
		fprintf(outfp, "     Salinity:         %f per mil\n", salinity);
		if (soundspeed > 0.0)
			fprintf(outfp, "     Soundspeed:       %f m/sec\n", soundspeed);
		fprintf(outfp, "     Depth:            %f m\n", depth);
		fprintf(outfp, "     pH:               %f\n", ph);
		fprintf(outfp, "Result:\n");
		fprintf(outfp, "     Sound absorption: %f dB/km\n", absorption);
		fprintf(outfp, "     Density:          %f kg/m3\n", density);
	}
	else {
		fprintf(outfp, "%f\n", absorption);
	}

	if (verbose >= 2) {
		fprintf(outfp, "\ndbg2  Program <%s> completed\n", program_name);
		fprintf(outfp, "dbg2  Ending status:\n");
		fprintf(outfp, "dbg2       status:  %d\n", status);
	}

	exit(error);
}
/*--------------------------------------------------------------------*/
